compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2776,7 +2776,7 @@ impl<'a> Parser<'a> {
         };
         // Try to parse the pattern `for ($PAT) in $EXPR`.
         let pat = match (
-            self.parse_pat_allow_top_alt(
+            self.parse_pat_allow_top_guard(
```

Comment:
I don't think this should allow guards to be parsed? This is an irrefutable context

Reply:
Despite the context being irrefutable, I believe it's better to parse them and emit a refutability error later. Doing it that way makes it easier to get a good error message (rather than just "unexpected token if..."), and prevents the situation where adding parentheses around the pattern substantially changes the error message.

It's also implied [by the RFC text](https://rust-lang.github.io/rfcs/3637-guard-patterns.html#in-assignment-like-contexts) that only `if let` and `while let` will be changed to use `PatternNoTopGuard`.

Reply:
If/when we allow `if let` pattern guards, then `x if let y = f(x)` is an example of an irrefutable pattern guard. Even if we don't allow it I expect users might write them, e.g.:
```rust
macro_rules! deref {
    ($p: pat) => { ref x if let $p = Deref::deref(x) }
}
fn foo(slice: &[Box<u32>]) {
    for deref!(x) in slice {
        // user expects x: &u32
    }
}
```